src/wallet/rpcdump.cpp

```diff
@@ -613,36 +608,60 @@ UniValue importwallet(const JSONRPCRequest& request)
-                   pwallet->WalletLogPrintf("Error importing script %s\n", vstr[0]);
-                   fGood = false;
-                   continue;
-               }
-               int64_t birth_time = DecodeDumpTime(vstr[1]);
-               if (birth_time > 0) {
-                   pwallet->m_script_metadata[id].nCreateTime = birth_time;
-                   nTimeBegin = std::min(nTimeBegin, birth_time);
-               }
+                std::vector<unsigned char> vData(ParseHex(vstr[0]));
+                CScript script = CScript(vData.begin(), vData.end());
+                int64_t birth_time = DecodeDumpTime(vstr[1]);
+                scripts.push_back(std::pair<CScript, int64_t>(script, birth_time));
             }
         }
         file.close();
+        for (unsigned int i = 0; i < keys.size(); ++i) {
+            const auto& key_tuple = keys[i];
+            uiInterface.ShowProgress("", std::max(50, std::min(75, (int)(((double)i / (double)(keys.size() + scripts.size())) * 100))), false);
+            const CKey& key = std::get<0>(key_tuple);
+            int64_t time = std::get<1>(key_tuple);
+            bool has_label = std::get<2>(key_tuple);
+            std::string label = std::get<3>(key_tuple);
+
+            CPubKey pubkey = key.GetPubKey();
+            assert(key.VerifyPubKey(pubkey));
+            CKeyID keyid = pubkey.GetID();
+            if (pwallet->HaveKey(keyid)) {
+                pwallet->WalletLogPrintf("Skipping import of %s (key already present)\n", EncodeDestination(keyid));
+                continue;
+            }
+            pwallet->WalletLogPrintf("Importing %s...\n", EncodeDestination(keyid));
+            if (!pwallet->AddKeyPubKey(key, pubkey)) {
+                fGood = false;
+                continue;
+            }
+            pwallet->mapKeyMetadata[keyid].nCreateTime = time;
+            if (has_label)
+                pwallet->SetAddressBook(keyid, label, "receive");
+            nTimeBegin = std::min(nTimeBegin, time);
+        }
+        for (unsigned int i = 0; i < scripts.size(); ++i) {
+            const auto& script_pair = scripts[i];
+            uiInterface.ShowProgress("", std::max(50, std::min(75, (int)(((double)(i + keys.size()) / (double)(keys.size() + scripts.size())) * 100))), false);
+            const CScript& script = script_pair.first;
+            int64_t time = script_pair.second;
+            CScriptID id(script);
+            if (pwallet->HaveCScript(id)) {
+                pwallet->WalletLogPrintf("Skipping import of %s (script already present)\n", HexStr(script));
+                continue;
+            }
+            if(!pwallet->AddCScript(script)) {
+                pwallet->WalletLogPrintf("Error importing script %s\n", HexStr(script));
+                fGood = false;
+                continue;
+            }
+            if (time > 0) {
+                pwallet->m_script_metadata[id].nCreateTime = time;
+                nTimeBegin = std::min(nTimeBegin, time);
```

Comment:
Shouldn't this be `birthtime`? `int64_t birth_time = DecodeDumpTime(vstr[1])` (because all script timestamps are 0 in the dump)

Reply:
No because `birth_time` just becomes `time` in this loop.

Reply:
Ok, but then what's causing the full rescan?

Just tried on `HEAD~2` and getting the same thing, so I'm blaming my own wallet. Just tried this again with a fresh wallet that only has one transaction. Now the rescan is a modest 10623 blocks (I guess that's a built in safety margin?).

 I also got a ton of `Skipping import of ... (script already present)` messages both my original weird wallet and on the fresh one. That doesn't make sense to me, because I'm importing into a brand new wallet.  But it's not new behavior.

Reply:
I'm not sure what might be causing that problem. However the `Skipping import` messages for scripts is expected since segwit scripts are in the dump, but already automatically imported when the keys were imported.